src/Compilers/Core/Portable/Symbols/Attributes/CommonAttributeData.cs

```diff
@@ -292,7 +292,25 @@ private ObsoleteAttributeData DecodeObsoleteAttribute()
                 }
             }
 
-            return new ObsoleteAttributeData(ObsoleteAttributeKind.Obsolete, message, isError);
+            TypedConstant diagnosticId = default;
+            TypedConstant urlFormat = default;
+            foreach (var pair in this.CommonNamedArguments)
+            {
+                switch (pair.Key)
+                {
+                    case "DiagnosticId":
+                        diagnosticId = pair.Value;
+                        break;
+                    case "UrlFormat":
+                        urlFormat = pair.Value;
+                        break;
+                    default:
+                        // unknown property specified on ObsoleteAttribute
+                        break;
+                }
+            }
+
+            return new ObsoleteAttributeData(ObsoleteAttributeKind.Obsolete, message, isError, (string?)diagnosticId.ValueInternal, (string?)urlFormat.ValueInternal);
```

Comment:
>(string?)diagnosticId.ValueInternal [](start = 95, length = 35)

How do we know that this cast is safe?

Reply:
We don't know this cast is safe or the casts above for the constructor arguments. Happy to change these casts to 'as' expressions. You can produce a crash today with a program like this:

```cs
using System;

namespace System
{
    public class ObsoleteAttribute : Attribute
    {
        public ObsoleteAttribute(string s1, string s2) { }
    }
}

public class C
{
    [Obsolete("a", "b")]
    public void M()
    {
    }
}
```

the crash:
```
System.AggregateException: One or more errors occurred. (Unable to cast object of type 'System.String' to type 'System.Boolean'.)

 ---> System.InvalidCastException: Unable to cast object of type 'System.String' to type 'System.Boolean'.

   at Microsoft.CodeAnalysis.AttributeData.DecodeObsoleteAttribute() in /_/src/Compilers/Core/Portable/Symbols/Attributes/CommonAttributeData.cs:line 283

   at Microsoft.CodeAnalysis.AttributeData.DecodeObsoleteAttribute(ObsoleteAttributeKind kind) in /_/src/Compilers/Core/Portable/Symbols/Attributes/CommonAttributeData.cs:line 252

   at Microsoft.CodeAnalysis.CSharp.Symbol.EarlyDecodeDeprecatedOrExperimentalOrObsoleteAttribute(EarlyDecodeWellKnownAttributeArguments`4& arguments, CSharpAttributeData& attributeData, ObsoleteAttributeData& obsoleteData) in /_/src/Compilers/CSharp/Portable/Symbols/Symbol_Attributes.cs:line 186

   at Microsoft.CodeAnalysis.CSharp.Symbols.SourceMethodSymbolWithAttributes.EarlyDecodeWellKnownAttribute(EarlyDecodeWellKnownAttributeArguments`4& arguments) in /_/src/Compilers/CSharp/Portable/Symbols/Source/SourceMethodSymbolWithAttributes.cs:line 340

   at Microsoft.CodeAnalysis.CSharp.Symbol.EarlyDecodeWellKnownAttributes(ImmutableArray`1 binders, ImmutableArray`1 boundAttributeTypes, ImmutableArray`1 attributesToBind, AttributeLocation symbolPart, CSharpAttributeData[] boundAttributesBuilder) in /_/src/Compilers/CSharp/Portable/Symbols/Symbol_Attributes.cs:line 579

   at Microsoft.CodeAnalysis.CSharp.Symbol.LoadAndValidateAttributes(OneOrMany`1 attributesSyntaxLists, CustomAttributesBag`1& lazyCustomAttributesBag, AttributeLocation symbolPart, Boolean earlyDecodingOnly, Binder binderOpt, Func`2 attributeMatchesOpt) in /_/src/Compilers/CSharp/Portable/Symbols/Symbol_Attributes.cs:line 362

   at Microsoft.CodeAnalysis.CSharp.Symbols.SourceMethodSymbolWithAttributes.GetAttributesBag(CustomAttributesBag`1& lazyCustomAttributesBag, Boolean forReturnType) in /_/src/Compilers/CSharp/Portable/Symbols/Source/SourceMethodSymbolWithAttributes.cs:line 279

   at Microsoft.CodeAnalysis.CSharp.Symbols.SourceMethodSymbolWithAttributes.GetAttributesBag() in /_/src/Compilers/CSharp/Portable/Symbols/Source/SourceMethodSymbolWithAttributes.cs:line 229

   at Microsoft.CodeAnalysis.CSharp.Symbols.SourceMethodSymbolWithAttributes.GetAttributes() in /_/src/Compilers/CSharp/Portable/Symbols/Source/SourceMethodSymbolWithAttributes.cs:line 293
```

Reply:
> We don't know this cast is safe or the casts above for the constructor arguments. Happy to change these casts to 'as' expressions.

The target constructor in the example doesn't match any of the signatures that we should recognize. We shouldn't even be calling this method. I am not sure the `as` cast is the answer for properties. It might be, but we should also verify that the names refer to real properties rather than fields, etc.

---
In reply to: [395362252](https://github.com/dotnet/roslyn/pull/42518#discussion_r395362252) [](ancestors = 395362252)

Reply:
> You can produce a crash today with a program like this ...

Please open an issue for the crash with existing code.

---
In reply to: [395379022](https://github.com/dotnet/roslyn/pull/42518#discussion_r395379022) [](ancestors = 395379022,395362252)

Reply:
Filed #42741